### ISSUE TITLE
MutableSharedFlow no works on iOS device

### ISSUE THREAD

**Issue by francismariano** [Kotlin/kotlinx.coroutines#2533]
Hello guys.

I have a KMM project with a common code using `implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:1.4.2-native-mt")`

NOTE: This code works like a charm on Android device.

In this project I have a Repository object.

```
object BluetoothRepositoryKmmImpl : BluetoothRepositoryKmm {
    .......
    private val _devices = MutableSharedFlow<Device>()
    override val devices = _devices.asSharedFlow()

    private val _advertisement = MutableSharedFlow<Advertisement>()
    val advertisement = _advertisement.asSharedFlow()

    override fun startScanLeDevice() {
        println("BluetoothRepositoryKmmImpl - startScanLeDevice222")
        scanJob = repositoryScope.launch {
            println("BluetoothRepositoryKmmImpl - startScanLeDevice:launch")
            Scanner().advertisements.collect {
                println("BluetoothRepositoryKmmImpl - startScanLeDevice:collect")
                _advertisement.emit(it)
                _devices.emit(it.devicePlatformFrom())
            }
        }.apply { invokeOnCompletion { println("BluetoothRepositoryKmmImpl - startScanLeDevice:invokeOnCompletation") } }
    }

    .....
}
```
BluetoothRepositoryKmmImpl - startScanLeDevice222
BluetoothRepositoryKmmImpl - startScanLeDevice:launch
BluetoothRepositoryKmmImpl - startScanLeDevice:collect
BluetoothRepositoryKmmImpl - startScanLeDevice:collect
BluetoothRepositoryKmmImpl - startScanLeDevice:collect
BluetoothRepositoryKmmImpl - startScanLeDevice:collect
BluetoothRepositoryKmmImpl - stopScanLeDevice
BluetoothRepositoryKmmImpl - startScanLeDevice:invokeOnCompletation

```
    fun deviceScanDevice() {
        view?.log("presenter - deviceScanDevice555")
        view?.showCancelButton()
        view?.clearDevices()
        scan = mainScope.launch {
            view?.log("presenter - deviceScanDevice:startScanLeDevice")
            bluetoothRepository.startScanLeDevice()
            try {
                view?.log("presenter - deviceScanDevice:startScanLeDevice:try")
                withTimeout(20000) {
                    view?.log("presenter - deviceScanDevice:withTimeout:withTimeout")
                    bluetoothRepository.devices.collect {
                        view?.log("presenter - deviceScanDevice:withTimeout:collect")
                        view?.showDevice(it)
                    }
                }
            } catch (e: TimeoutCancellationException) {
                view?.log("presenter - deviceScanDevice:catch : ${e.message}")
                view?.showUpdateButton()
            }
        }.apply {
            invokeOnCompletion {
                bluetoothRepository.stopScanLeDevice()
                view?.log("presenter - deviceScanDevice555:invokeOnCompletation")
            }
        }
    }
```

DeviceScanViewController:log = presenter - onViewAttached
DeviceScanViewController:log = presenter - deviceScanDevice555
DeviceScanViewController:log = presenter - deviceScanDevice:startScanLeDevice
DeviceScanViewController:log = presenter - deviceScanDevice:startScanLeDevice:try
DeviceScanViewController:log = presenter - deviceScanDevice:withTimeout:withTimeout
DeviceScanViewController:log = presenter - deviceScanDevice:catch : Timed out waiting for 20000 ms
DeviceScanViewController:log = presenter - deviceScanDevice555:invokeOnCompletation

In summary, the startScanLeDevice function collect the data but deviceScanDevice function does not collect data from bluetoothRepository.devices. 

It is a bug? How can I debug this?

Thank you.

**Comment by francismariano** [comment]
The problem was the @ThreadLocal annotation in BluetoothRepositoryKmmImpl object. Removing the annotation and using atomics solves the problem.
